fggs/semirings.py

```diff
@@ -144,22 +148,25 @@ def from_int(self, n):
         n = torch.as_tensor(n, dtype=self.dtype, device=self.device)
         return torch.log(n)
     
-    add = staticmethod(torch.logaddexp) # type: ignore
+    @staticmethod
+    def add(x, y):
+        return x.logaddexp(y)
+
     sum = staticmethod(torch.logsumexp) # type: ignore
-    
+
     @staticmethod
-    def sub(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+    def sub(x, y):
         # If x <= y, return -inf
         # If x ≈ y, log(exp(x) - exp(y)) = x + log(1 - exp(y-x)) = x + log(-expm1(y-x))
         # If x >> y, log(exp(x) - exp(y)) = x + log(1 - exp(y-x)) = x + log1p(-exp(y-x))
-        d = y - x
-        return torch.where(d < -1, # type: ignore
-                           x + torch.log1p(-torch.exp(d)), # type: ignore
-                           x + torch.log(-torch.expm1(d))).nan_to_num(nan=-inf, neginf=-inf, posinf=inf)
-    
+        d = y.sub(x)
```

Comment:
Are the methods faster than the operators?

Reply:
To work with both Tensor and EmbeddedTensors, the code must invoke methods (such as `t.exp()`) instead of functions (such as `torch.exp(t)`). But operators are methods (for instance, `y-x` means `y.__sub__(x)`).

Initially I was avoiding operator overloading because I was so unfamiliar with it. Now I think I can change `sub` to `__sub__` (and invoke it using operator syntax) if desired.